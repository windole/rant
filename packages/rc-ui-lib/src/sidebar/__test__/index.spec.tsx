import React from 'react';
import { fireEvent, render, cleanup } from '@testing-library/react';
import { Sidebar } from '..';

describe('Sidebar', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('should emit change event when active item changed', async () => {
    const onChange = jest.fn();

    const { container } = render(
      <Sidebar onChange={onChange}>
        <Sidebar.Item title="标签名" />
        <Sidebar.Item title="标签名" />
        <Sidebar.Item title="标签名" />
      </Sidebar>,
    );
    const wrapper = container.querySelectorAll('.rc-sidebar-item');
    await fireEvent.click(wrapper[0]);
    expect(onChange).toHaveBeenCalledTimes(0);
    await fireEvent.click(wrapper[1]);
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('should emit click event when SidebarItem is clicked', async () => {
    const onClick = jest.fn();

    const { container } = render(
      <Sidebar>
        <Sidebar.Item onClick={onClick} title="标签名">
          Text
        </Sidebar.Item>
      </Sidebar>,
    );
    const wrapper = container.querySelectorAll('.rc-sidebar-item');
    await fireEvent.click(wrapper[0]);
    expect(onClick).toHaveBeenCalledWith(0);
  });

  it('should not emit change event when disabled SidebarItem is clicked', async () => {
    const onChange = jest.fn();

    const { container } = render(
      <Sidebar onChange={onChange}>
        <Sidebar.Item title="标签名" />
        <Sidebar.Item title="标签名" disabled />
        <Sidebar.Item title="标签名" />
      </Sidebar>,
    );
    const wrapper = container.querySelectorAll('.rc-sidebar-item');
    await fireEvent.click(wrapper[1]);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should render correctly when using badge prop in SidebarItem', async () => {
    const { container } = render(
      <Sidebar>
        <Sidebar.Item title="标签名" disabled />
        <Sidebar.Item badge="5" title="标签名" />
        <Sidebar.Item badge="20" title="标签名" />
      </Sidebar>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render incorrectly when SidebarItem not parent', async () => {
    const spyConsole: jest.SpyInstance = jest.spyOn(console, 'error');
    spyConsole.mockImplementation(() => {
      return null;
    });
    // eslint-disable-next-line react/no-children-prop
    const { container } = render(<Sidebar.Item />);

    expect(container).toMatchSnapshot();
    expect(spyConsole).toHaveBeenCalledWith(
      '[rc-ui-lib] <SidebarItem> must be a child component of <Sidebar>.',
    );
    spyConsole.mockRestore();
  });
});
