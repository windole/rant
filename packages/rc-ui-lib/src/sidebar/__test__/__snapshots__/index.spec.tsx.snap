// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render correctly when using badge prop in SidebarItem 1`] = `
<div>
  <div
    class="rc-sidebar"
    role="tablist"
  >
    <div
      aria-selected="true"
      class="rc-sidebar-item rc-sidebar-item--select rc-sidebar-item--disabled"
      role="tab"
    >
      <div
        class="rc-badge__wrapper rc-sidebar-item__text"
      >
        标签名
      </div>
    </div>
    <div
      aria-selected="false"
      class="rc-sidebar-item"
      role="tab"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-sidebar-item__text"
      >
        标签名
        <div
          class="rc-badge rc-badge--fixed"
        >
          5
        </div>
      </div>
    </div>
    <div
      aria-selected="false"
      class="rc-sidebar-item"
      role="tab"
      tabindex="0"
    >
      <div
        class="rc-badge__wrapper rc-sidebar-item__text"
      >
        标签名
        <div
          class="rc-badge rc-badge--fixed"
        >
          20
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sidebar should render incorrectly when SidebarItem not parent 1`] = `<div />`;
